route/v1/types.go

```diff
@@ -350,6 +353,8 @@ type RouteStatus struct {
 	// ingress describes the places where the route may be exposed. The list of
 	// ingress points may contain duplicate Host or RouterName values. Routes
 	// are considered live once they are `Ready`
+	// +listType=map
+	// +listMapKey=routerName
```

Comment:
CI is complaining with 
```
spec.validation.openAPIV3Schema.properties[status].properties[ingress].items.properties[routerName].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property",
```
The comment on the `RouterName` field does mention that `this value is required`, but the validation is not enforced. I believe we need to remove the `omitempty` json tag?

cc: @JoelSpeed

Reply:
This _could_ be a breaking change, but, IIRC, routers are actually served by openshift-apiserver, so, do we know if the openshift-apiserver is enforcing that validation already, if it is, I think we can move towards making it required

@Miciah do you know?

If the field isn't actually required then we will have to rethink this change

Reply:
Looks like no, there's no validation on the route status, https://github.com/openshift/library-go/blob/8bb8fe6cc56df74b00c0ee71420d4038d4a2281c/pkg/route/validation/validation.go#L226-L235

Reply:
Yeah, this seems like a breaking change if the route status is getting updated somewhere without this field.  I'll have to defer to @Miciah to confirm if this could be an issue. 


Having said that, I'm wondering why CI didn't catch the missing SSA tags earlier and how renaming (removing) a feature gate is causing it.

Reply:
> Having said that, I'm wondering why CI didn't catch the missing SSA tags earlier and how renaming (removing) a feature gate is causing it.

Are the two APIs in the same sets?

Reply:
@JoelSpeed, Chirag and I discussed this issue today.  

A managed router deployment always has a canonical name set, so the router will set `routerName` on any status ingress entry that the managed router adds.  The only risk with making `routerName` required would be if some third-party ingress-controller implementation uses the Route API (this is unlikely) and sets a status entry without specifying `routerName`, or if the cluster has a manually configured router (this is more likely) that does not have a canonical name set.  

One option we discussed was setting a default value.  Managed routers always have a canonical name of the form "router-", so as long as the default value for `routerName` does not start with "router-", it will not conflict with managed routers.  Then the only risk is if some third-party software uses the Route API, sets status ingress entries with empty `routerName`, *and* breaks when its status entry's `routerName` is nonempty.  I think this scenario is extremely unlikely, it would arguably be a defect in the third-party software, and after all, we did say all along that `routerName` was required (even if we didn't enforce it).  

Meanwhile, the critical thing is that setting a default value avoids the situation where marking the field as required causes updates on an existing object with empty `routerName` (if such an object exists) to fail—right?

So what do you think of marking the field as required, with some default value?

Reply:
@JoelSpeed, in any case, if we cannot come up with a quick resolution to this issue, can we defer adding `listType` to a separate PR and override the failing CI tests for this PR?

Reply:
I don't think you can have a required field with a default, defaulting is applied after the validation has already run.

Given we are confident that our controllers set the value correctly, I'd be happy to merge as moving it to required. We are slightly more flexible in status than we are in spec, as we don't expect external consumers.

Also, now that we are in 4.18, we should have ratcheting validation by default in the API server, as such, existing values should remain and only updates to the specific values should break.

This will need to be tested though, I have a todo for adding schema upgrade tests which I hope to look at this week/next week

Reply:
> Also, now that we are in 4.18, we should have ratcheting validation by default in the API server, as such, existing values should remain and only updates to the specific values should break.

Generally speaking, this is excellent news about ratcheting validation.  However, using ratcheting here would prevent us from backporting the PR to 4.16 (which introduced the erroneously named featuregate).  For that reason, it would be preferable to split the change to add tags into a separate PR, and keep this PR focused on the removal of the featuregate.

Reply:
Ok, then lets work out a separate path that can be backported, and look at the feature gate removal now. I think lets change this just to atomic and we can follow up to make it a map for 4.18 onwards

Reply:
Okay, I've changed the listType to atomic as of now. We can follow up on this as part of a separate ticket.